### DISCUSSION TITLE
Storage: image resize

### DISCUSSION THREAD

**Discussion by joshnuss** [supabase/supabase#1407] [Feature Requests] [answered]
### Update: [image resizing is now available](https://github.com/orgs/supabase/discussions/1407#discussioncomment-4391518)

---

It would be helpful if we could scale images/videos with query params.

Similar to how Cloudinary and other storage providers do it. Example: https://cloudinary.com/cookbook/resize_an_image

One way to do this is by placing a proxy in front of the storage to cache and serve scaled assets.

**Comment by inian** [comment]
Hey @joshnuss, this is on the roadmap but some blockers before we implement it.

First, we will probably launch a CDN integration for assets in storage which is the perfect "cache" for these use cases. 

There is also work that needs to be done on making storage-api multi-tenant which will let us run transformations like these on beefier systems. The current resources we have provisioned for our free tier may not have enough power to do CPU / GPU intensive tasks like these at a reasonable scale. 

Stay tuned for more updates :)

**Reply by joshnuss** [reply]
Awesome! 
Thank @inian, glad to hear you're working on it

**Reply by LyMc** [reply]
@inian, now that CDN is integrated with Supabase, do you have an estimation for image transformations?

**Reply by eMeRiKa13** [reply]
Nice question @LyMc :) Waiting for the answer as well.

**Reply by eMeRiKa13** [reply]
In the meantime, do you have any other solutions to recommend to handle image resizing?

**Reply by LyMc** [reply]
I use next image for now, it resize the image on request, but it can cache the result, so there is a delay only on first request.

**Reply by eMeRiKa13** [reply]
@LyMc thanks for your answer.

**Reply by inian** [reply]
Hey everyone, this is the next feature for storage we are going to be working on. We are planning our internal roadmap now, but this should be launched in March / April 2022 roughly

**Reply by LyMc** [reply]
Thanks for the update, looks good, waiting for this feature 🤩

**Reply by eMeRiKa13** [reply]
Awesome, thanks for the information.

**Reply by bradgreens** [reply]
> We are planning our internal roadmap now, but this should be launched in March / April 2022 roughly

How is the progress on this roadmap feature going? Very interested as well.

**Reply by inian** [reply]
This got de-prioritised for [this launch week](https://supabase.com/blog/2022/03/25/supabase-launch-week-four), since we were working on another feature which loads of folks were asking for 👀  Updated timeline is sometime this June

**Reply by zidkim** [reply]
@inian appreciate the update.

**Reply by peachp** [reply]
@inian thanks for the update. June (even July still) would be good timing for us.
A bit background info: we are a photos-heavy business, currently migrating from Airtable/Nuxt/DigitalOcean to Supabase/SvelteKit and it would be great to use Supabase Storage for storing, resizing and serving photos via CDN, as otherwise I'd have to keep using DO or switch to Cloudflare Images or similar.

**Reply by irreal** [reply]
Hey, any update on the timeline for resizing images? We need that feature. if it's coming relatively soon, we won't waste time integrating a 3rd party service now.

**Reply by obrunsmann** [reply]
Same here... about to implement by my own but would be nice to have any idea about the current timeline 👀

**Reply by emileond** [reply]
+1

**Reply by navalabe** [reply]
+1

**Reply by zidkim** [reply]
I've been told a few days ago we should expect a release around 1-2 months from now.

**Reply by madeleineostoja** [reply]
This would be incredible. I've been pulling my hair out trying to find a scalable solution to image resizing (imgix gets very expensive very quickly if you have a high asset / low bandwidth scenario, imageproxy looks painful). Would love to get an update from the Supabase team on the rough roadmap for this one to decide whether I keep paying through the teeth for 3rd parties in the meantime or invest into building my own proxy infra

**Reply by inian** [reply]
Hi all, this is the top priority for the storage team now and we have started work on this already. We will be launching with image resizes first. Let me know what other transformations you would like to see!

**Reply by madeleineostoja** [reply]
Woohoo! Format conversion and a quality filter would be excellent too, basically just the basics needed for good responsive images (webp/avif srcsets with slight compression)

**Reply by zakelfassi** [reply]
An object background remover? One can dream (plus, the next best API option in the market costs a leg, so any offering will be a good competing offering)

**Reply by christianfl** [reply]
I'd dream of a possibility to request videos with different resolutions. And some video on-the-fly streaming to not serve the whole video would be insane... Can understand if that would be too much work, though.

**Reply by hjaber** [reply]
For images, it would be great to have the basics well executed to start out, resizing and compression similar to Cloudfare image resize workers. 

Strip all metadata, allow for lossless or lossy format. Similar to nextjs image, allow for lossy format. An use case would be to reduce quality of smaller images like thumbnails. 

Allow for multiple size variants, Cloudfare images gives you the ability to serve different sizes which is part of modern web responsive design. 

Custom domains would be a future nice to have feature as it is possible with Cloudfare workers.

**Reply by mdarche** [reply]
+1 for custom domains - this would be an absolute game changer. Thank you guys at Supabase for all of your great work!

**Reply by BhaskaranR** [reply]
@inian  sorry to bother. do we have a date for beta?

**Reply by inian** [reply]
Hi everyone, we can start rolling out early access to this feature as early as next month. Please reach out to growth@supabase.io if you are interested in this!

**Reply by madeleineostoja** [reply]
Emailed! Do you have an idea of pricing yet?

**Reply by inian** [reply]
We are still working on the pricing now, nothing concrete yet.

**Reply by madeleineostoja** [reply]
Awesome. It would be wonderful if it was based on storage use rather than number of objects like Imgix. Imgix really screws you if you have a high volume / low usage scenario, which I imagine is the case for a lot of early stage apps

**Reply by inian** [reply]
Ah I see you want a pricing based on number of actual transformations done instead of number of images in the bucket even though they may not be actually transformed?

**Reply by peachp** [reply]
@inian we have the same situation like @madeleineostoja : we have ~10,000 images, but quite low traffic (100k monthly views), so paying for just storing 10,000 images all the time from which only a fraction is accessed in a give month would be suboptimal for us.

By the sound of it, Transformations will be charged extra on top of the Storage allowances we already get with the Pro plan? I mean I don't mind paying a few extra bucks a month for a good service, as long as it's charged fairly based on what we actually use (see above).

**Reply by hjaber** [reply]
> Ah I see you want a pricing based on number of actual transformations done instead of number of images in the bucket even though they may not be actually transformed?

It is very difficult to provide a one size fits all best pricing for each person's use case. 

Maybe it would be worthwhile to ask the community what they are looking for. 

Cloudfare recently introduced Cloudfare images and it is the best I've seen for one size fits all in terms of pricing. $5 per 100,000 images and $1 per 100,000 images delivered with no egress or charge for resizing optimization (variants). It works well for projects that only have 10,000 images and lots of traffic as well as 500,000 with low traffic. It also automatically serves images to Cloudfare's CDN.

**Reply by madeleineostoja** [reply]
I also love cloudflare’s pricing model (charge low prices for both types of usage, so it scales fairly however it’s combined). We actually built the first spike of our app on Cloudflare Images purely because of the unit economics it would have given us, before moving to supabase storage + imgix because the platform itself was too buggy and support was nonexistent.

Supabase already has pricing established for storage itself, which is a close enough substitute for number of images, so I feel like adding a transformations requests dimensions to the existing pricing could work well. Since we already get storage for “free” (ie already paying for it) it will work out cheaper than Imgix etc storing copies of every image no matter what

**Reply by inian** [reply]
Image resizing with a smart cdn is now available! https://supabase.com/blog/storage-image-resizing-smart-cdn

Let me know what you folks think!

**Reply by madeleineostoja** [reply]
Thank you @inian! Said it in a seperate comment below, but we probably wont migrate to Supabase from imgix for our image optimisation pipeline until at least format conversion to webp/avif is supported, and ideally a quality filter as well. Even with resizing, full quality jpegs are too chonky for performant responsive images. And transforming source images before upload serverside isn't an option for us because we want to give users the option to enable downloads of their original full res images (we run a photography portfolio platform).

Super exciting to see it happening though! I like the pricing you landed on, makes it competitive with imgix et al and I don't have to worry about bandwidth/request caps

**Reply by inian** [reply]
webp is coming really soon. Resizing took so long cos we had to build a lot of the underlying infrastructure, now thats in place, adding new transforms shouldnt be hard!

**Comment by inian** [comment, marked as answer]
Image resizing with a smart cdn is now available! https://supabase.com/blog/storage-image-resizing-smart-cdn

**Reply by wiesson** [reply]
@inian - is there a hard limit for the longest side of the image? I had a couple of images images where I could transform them (pro plan)

```{
"statusCode": "422",
"error": "ImageProcessingError",
"message": "Invalid source image"
}
```

I resized the image and re uploaded them and it looks like the longest side needs to be <= 2560 for webp, but not for jpeg. I think that 2560px should be ok, but if I take a picture from my phone and upload it has larger dimensions.

**Comment by peachp** [comment]
@inian sounds good! I'll start using it ASAP for our website rebuild. Question regarding pricing. It sounds like it could be a deal breaker for us. In the blog post it says:
_"You can transform 100 images per project, at no additional cost. Beyond this, it is $5 per additional 1000 images transformed."_

We have around 10,000 images. If we assume all 10,000 images will be uploaded to Storage in one go and then every image will be requested every month at least once, does it mean we'll have to pay $50 every month? Or only once (in the first month)?

Thanks!

**Reply by madeleineostoja** [reply]
From the emails I've had with Supabase it's a monthly charge based on the number of images requested (ie: transformed) each month. The pricing scales quite well compared to imgix et al

**Reply by peachp** [reply]
Well, in real world I guess not all of our 10,000 images are requested every month, but it could be a big part of it... Hm, maybe I will transform them myself after all...

**Comment by madeleineostoja** [comment]
Awesome! For anyone that missed it in the announcement post, head here to vote for more transforms: https://github.com/supabase/supabase/discussions/10918

IMO not having format conversion and a quality filter are dealbreakers for any image optimisation feature, and will stop us from adopting this over imgix until they're implemented (format conversion to at least webp in particular).

**Reply by inian** [reply]
webp is coming really soon. Resizing took so long cos we had to build a lot of the underlying infrastructure, now thats in place, adding new transforms shouldn't be hard!
